src/dune_rules/cinaps.ml

```diff
@@ -69,6 +59,28 @@ let gen_rules sctx t ~dir ~scope =
     Super_context.resolve_program sctx ~dir ~loc:(Some loc) name
       ~hint:"opam install cinaps"
   in
+  let cinaps_dir =
+    let stamp =
+      let digest =
+        if cinapsed_files = [] then
+          Digest.generic (t.loc, t.libraries, t.preprocess, t.preprocessor_deps)
```

Comment:
Why the special case?

Reply:
For super special cases like this:

```ocaml
(cinaps (files *.does_not_exist))
(cinaps (files *.also_does_not_exist))
```

Where both `files` fields match nothing.

Reply:
The problem is more general. You have the same issue as soon as you have two identical cinaps stanzas. The way to solve it in general is to do the same as for ppx drivers and generate shared cinaps drivers on demand.

Reply:
> You have the same issue as soon as you have two identical cinaps stanzas.

If you have two identical cinaps stanzas, then they will match the same files. That's an error anyway.

>  The way to solve it in general is to do the same as for ppx drivers and generate shared cinaps drivers on demand.

I'll look into that. Happy to do it if it isn't a huge hassle and useful elsewhere.

Reply:
> If you have two identical cinaps stanzas, then they will match the same files. That's an error anyway.

I don't think so. You will still end up with the same stamp and so you'll get two rules trying to produce the same cinaps.exe.

Reply:
> You will still end up with the same stamp and so you'll get two rules trying to produce the same cinaps.exe.

I agree that this will happen. But what I'm saying is that even if we assign different stamps, it will fail later anyway. So it's just a matter of error message.

Reply:
> > The way to solve it in general is to do the same as for ppx drivers and generate shared cinaps drivers on demand.

> I'll look into that. Happy to do it if it isn't a huge hassle and useful elsewhere.

Actually, never mind about that. I realised that this won't work here as we need one custom program per `cinaps` stanza. It would have worked with the old design of cinaps that  was using the toplevel, but not the new one.

> I agree that this will happen. But what I'm saying is that even if we assign different stamps, it will fail later anyway due to the duplicate targets. So it's just a matter of error message.

The duplicate targets check should be a last resort though, as it might be difficult for a user to understand what they did wrong and how to move forward.

In any case, the PR in its current form is still an improvement over the status quo. So no need to block on this point.